File: 242. Valid Anagram.cpp
/*
questions:
Given two strings s and t , write a function to determine if t is an anagram of s.

Example 1:

Input: s = "anagram", t = "nagaram"
Output: true
Example 2:

Input: s = "rat", t = "car"
Output: false
*/
/*. sol1: use map */
class Solution {
public:
    bool isAnagram(string s, string t) {
        if(s.length() != t.length()) return false;
        unordered_map<char,int> MS;
        unordered_map<char,int> MT;
        for(int i=0;i<26;i++){
            MS['a'+i] = i;
            MT['a'+i] = i;
        }
        for(int i=0;i<s.length();i++){
            MS[s[i]]++;
            MT[t[i]]++;
        }
        return (MS == MT);
    }
};
/* sol2.use array to count */
class Solution {
public:
    bool isAnagram(string s, string t) {
        if(s.length() != t.length()) return false;
        vector<int> Count(26,0);

        for(int i=0;i<s.length();i++)
            Count[s[i]-'a']++;
        for(int i=0;i<s.length();i++)
            Count[t[i]-'a']--;
        for(int i=0;i<26;i++)
            if(Count[i]!=0) return false;
        return true;
    }
};

File: 242. Valid Anagram.py
#questions:
'''
Given two strings s and t , write a function to determine if t is an anagram of s.
Example 1:
Input: s = "anagram", t = "nagaram"
Output: true
Example 2:
Input: s = "rat", t = "car"
Output: false
'''

class Solution:
    def isAnagram(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        # sol1: dict
        A,B = {},{}
        for item in s:
            A[item] = A.get(item,0)+1
        for item in t:
            B[item] = B.get(item,0)+1
        return A==B
        
        # sol2: array
        A,B = [0]*26,[0]*26
        for item in s:
            A[ord(item)-ord('a')] +=1
        for item in t:
            B[ord(item)-ord('a')] +=1
        return A==B
        
        # sol3: string
        return sorted(s)==sorted(t)